### PR TITLE
Add test for inter-process services

### PR DESCRIPTION
**Pull request by jacquelinekay**
Work in progress (test fails for 1 middleware).

Adding a test for request/response between a client and a server in different processes, using nosetest. Lo and behold, the test times out for Connext Static but not Opensplice or Connext Dynamic.

@wjwwood this is for debugging the use_rcl branch of rclcpp, it seems to me that the problem of hanging services is in rcl, not rclcpp.

### CODE REVIEW

**Comment by wjwwood** [comment]
Hmm ok, I'll try this out locally too.

**Comment by wjwwood** [comment]
I can't run any of the service tests with this:

```
test 38
    Start 38: test_services__rmw_opensplice_cpp

38: Test command: /usr/local/bin/python3 "-u" "/Users/william/ros2_ws/install_isolated/ament_cmake_test/share/ament_cmake_test/cmake/run_test.py" "/Users/william/ros2_ws/build_isolated/rcl/test_results/rcl/test_services__rmw_opensplice_cpp.xunit.xml" "--output-file" "/Users/william/ros2_ws/build_isolated/rcl/ament_cmake_nose/test_services__rmw_opensplice_cpp.txt" "--command" "/usr/local/bin/nosetests" "/Users/william/ros2_ws/build_isolated/rcl/test/test_services__rmw_opensplice_cpp_RelWithDebInfo.py" "--with-xunit" "--xunit-file=/Users/william/ros2_ws/build_isolated/rcl/test_results/rcl/test_services__rmw_opensplice_cpp.xunit.xml" "--xunit-testsuite-name=rcl.nosetests"
38: Test timeout computed to be: 15
38: -- run_test.py: invoking following command in '/Users/william/ros2_ws/src/ros2/rcl/rcl':
38:  - /usr/local/bin/nosetests /Users/william/ros2_ws/build_isolated/rcl/test/test_services__rmw_opensplice_cpp_RelWithDebInfo.py --with-xunit --xunit-file=/Users/william/ros2_ws/build_isolated/rcl/test_results/rcl/test_services__rmw_opensplice_cpp.xunit.xml --xunit-testsuite-name=rcl.nosetests
38: E
38: ======================================================================
38: ERROR: Failure: ImportError (No module named 'test_services__rmw_opensplice_cpp_RelWithDebInfo')
38: ----------------------------------------------------------------------
38: Traceback (most recent call last):
38:   File "/usr/local/lib/python3.5/site-packages/nose/failure.py", line 39, in runTest
38:     raise self.exc_val.with_traceback(self.tb)
38:   File "/usr/local/lib/python3.5/site-packages/nose/loader.py", line 418, in loadTestsFromName
38:     addr.filename, addr.module)
38:   File "/usr/local/lib/python3.5/site-packages/nose/importer.py", line 47, in importFromPath
38:     return self.importFromDir(dir_path, fqname)
38:   File "/usr/local/lib/python3.5/site-packages/nose/importer.py", line 79, in importFromDir
38:     fh, filename, desc = find_module(part, path)
38:   File "/usr/local/Cellar/python3/3.5.1/Frameworks/Python.framework/Versions/3.5/lib/python3.5/imp.py", line 296, in find_module
38:     raise ImportError(_ERR_MSG.format(name), name=name)
38: ImportError: No module named 'test_services__rmw_opensplice_cpp_RelWithDebInfo'
38:
38: ----------------------------------------------------------------------
38: Ran 1 test in 0.001s
38:
38: FAILED (errors=1)
38: -- run_test.py: return code 1
38: -- run_test.py: verify result file '/Users/william/ros2_ws/build_isolated/rcl/test_results/rcl/test_services__rmw_opensplice_cpp.xunit.xml'
1/1 Test #38: test_services__rmw_opensplice_cpp ...***Failed    0.59 sec
```

**Comment by jacquelinekay** [comment]
@wjwwood I think the problem with running services tests was fixed in 71bc5b8 and 33b3b5e. But the services test don't pass :/ I'm going bang my head against this problem today.

**Comment by wjwwood** [comment]
@jacquelinekay I pushed a change I think you should pull (otherwise it might cause nasty conflicts with your future changes).

**Comment by jacquelinekay** [comment]
Thanks for letting me know. I've been playing around with rosidl/rclpy tweaks to get rid of the hardcoded identifier map, so I didn't have any pending conflicting changes.

**Comment by jacquelinekay** [comment]
I think all the tests are fixed (needed to change the default QoS settings and fix my exit handler). I made a branch called `use_rcl` that is identical to this and launched CI to kill 2 birds with one Jenkins job. If the services tests in rcl pass and the services tests in test_rclcpp fail then rclcpp needs more work. If it all passes then I think the rcl/rclcpp refactor is good to merge.

**Comment by jacquelinekay** [comment]
http://ci.ros2.org/job/ci_linux/1195/
http://ci.ros2.org/job/ci_osx/971/
http://ci.ros2.org/job/ci_windows/1250/

**Comment by wjwwood** [comment]
+1, but the cmake linter fails for me.

**Comment by wjwwood** [comment]
```
test 4
    Start 4: lint_cmake

4: Test command: /usr/local/bin/python3 "-u" "/Users/william/ros2_ws/install_isolated/ament_cmake_test/share/ament_cmake_test/cmake/run_test.py" "/Users/william/ros2_ws/build_isolated/rcl/test_results/rcl/lint_cmake.xunit.xml" "--output-file" "/Users/william/ros2_ws/build_isolated/rcl/ament_lint_cmake/lint_cmake.txt" "--command" "/Users/william/ros2_ws/install_isolated/ament_lint_cmake/bin/ament_lint_cmake" "--xunit-file" "/Users/william/ros2_ws/build_isolated/rcl/test_results/rcl/lint_cmake.xunit.xml"
4: Test timeout computed to be: 60
4: -- run_test.py: invoking following command in '/Users/william/ros2_ws/src/ros2/rcl/rcl':
4:  - /Users/william/ros2_ws/install_isolated/ament_lint_cmake/bin/ament_lint_cmake --xunit-file /Users/william/ros2_ws/build_isolated/rcl/test_results/rcl/lint_cmake.xunit.xml
4: test/rcl_add_custom_executable.cmake:28: Lines should be <= 80 characters long [linelength]
4:
4:
4: 1 errors
4: -- run_test.py: return code 1
4: -- run_test.py: verify result file '/Users/william/ros2_ws/build_isolated/rcl/test_results/rcl/lint_cmake.xunit.xml'
4/5 Test #4: lint_cmake .......................***Failed    0.46 sec
```

**Comment by wjwwood** [comment]
Ah, I guess you didn't push here: https://github.com/ros2/rcl/compare/more_tests...use_rcl

**Comment by jacquelinekay** [comment]
Fixed in 256c86e46b4397ec3bafe57058b7b97905627aae

**Comment by jacquelinekay** [comment]
yeah

**Comment by wjwwood** [comment]
Don't forget to squash, there's several small fixup commits here.

**Comment by wjwwood** [comment]
Also, did the CI finish?

**Comment by jacquelinekay** [comment]
The squash Github button is my friend, though it makes it look like the PR didn't get squashed in the web interface.

CI hasn't finished yet but I took your local testing as OK to merge. I'll deal with any regressions on Windows if they come up (I assume you tested on OSX).

**Comment by wjwwood** [comment]
Ok, squash button is good.

@jacquelinekay I tested some of the things on OS X, but not the latest changes. I'll just wait on the CI since it's already almost done.

File: rcl/test/rcl_add_custom_executable.cmake [hunk 1]
```diff
@@ -0,0 +1,71 @@
...
+  if(_ARG_APPEND_ENV)
+    set(_ARG_APPEND_ENV "APPEND_ENV" ${_ARG_APPEND_ENV})
+  endif()
+  if(_ARG_APPEND_LIBRARY_DIRS)
+    set(_ARG_APPEND_LIBRARY_DIRS "APPEND_LIBRARY_DIRS" ${_ARG_APPEND_LIBRARY_DIRS})
+  endif()
+
+  add_executable(${target} ${_ARG_SRCS})
+
+  if(TARGET ${target})
```
Comment (dirk-thomas) [inline rcl/test/rcl_add_custom_executable.cmake]: This test doesn't make sense since the target is unconditionally defined two lines above.

**Pull request merged**